### ISSUE TITLE
Add Execution Test Infrastructure

### ISSUE THREAD

**Issue by mgrover1** [ProjectPythiaTutorials/radar-cookbook#6]
Add test to ensure that notebooks execute - similar to the [Pythia Foundations](https://github.com/ProjectPythia/pythia-foundations) repo.

@brian-rose would you be willing to do this? Otherwise I can take care of it!

**Comment by brian-rose** [comment]
I'll take a stab at it.

**Comment by brian-rose** [comment]
First steps added in #8 and #9. 

The idea is to do all the building and previewing of PRs right here on GitHub in order to avoid complications with configuring a third-party service like Netlify, so that every new Cookbook can have the testing and preview capability immediately out of the box.

When a new PR is opened, a message is generated that gives a link to where the previews will be deployed, which is a subdirectory of wherever the Cookbook itself is hosted (on gh-pages) labeled with `.../_preview/[PR number]/`

Each new update to the PR branch will trigger a new build of book, which then gets deployed to the preview location.

When the PR is merged, the built site is deployed at the base URL and everything at `.../_preview/[PR number]` is deleted.

Currently still working out some glitches with all the above, as seen in #10

- [ ] The PR number is not getting appended to the preview URL
- [ ] There is no "delete the preview" action yet
- [ ] There is currently a redundant build-and-deploy action happening in parallel with the preview, this should be replaced with an action that deploys the already-built site upon merge
- [ ] Once this is all working, the build action can be put on a cron schedule for regular health checking
- [ ] We can also add a link-checker action if we want

**Comment by brian-rose** [comment]
#13 now has a working version of this concept. 

I ended up using workflow code very similar to what we have in [Foundations](https://github.com/ProjectPythia/pythia-foundations/blob/main/.github/workflows/preview-book.yaml), so it should look and feel familiar. But this deploys to gh-pages instead of Netlify.

Still a work in progress, but working!

@mgrover1 @kmpaul and anyone else who's interested, take a look at #13 to see what I mean.

**Comment by mgrover1** [comment]
Thanks @brian-rose !! This is awesome work!

**Comment by brian-rose** [comment]
Updates:
- Builds are accelerated now by using mambaforge to setup the conda environment and caching the environment for reuse
- There is a link-checker action that runs on PRs
- The previews are actually deleted automatically when a PR is merged.

That last point may be a fatal flaw in this approach. It works great when dealing with a single PR: the preview remains visible until the PR is merged and the site is re-rendered. But unfortunately merging a PR will deletes not just the preview associated that PR but also any others that are currently open.

**Comment by brian-rose** [comment]
I found a good solution to the multiple-preview problem, described over at https://github.com/brian-rose/ClimateLaboratoryBook/issues/74. I will implement and test this here.

Once that's all working, I think all that's left is setting up a schedule for regular builds, and getting nice status badges.

**Comment by brian-rose** [comment]
This is now done with #27 and tested. It works!

Previews are now persistent while a PR remains open, but get deleted when the PR is either closed or merged. Other previews are unaffected.

**Comment by brian-rose** [comment]
@mgrover1 I think this is ready for more eyes and test-drives. If you have a chance to open some new PRs on the radar cookbook, you can see the preview machinery in action. I'm pretty happy with it.

Still to do:

- [ ] Set up a cron schedule for regular build and link-check
- [ ] Put status badges for those builds on the README page
- [ ] Migrate all this infrastructure upstream to the Cookbook template

**Comment by brian-rose** [comment]
Scheduling and badges added in #30 and #31 respectively. Assuming the scheduled nightly build works as expected, I think this is done.

**Comment by brian-rose** [comment]
Not critical, but the scheduled run of `build_book.yaml` causes the `deploy_preview.yaml` workflow to run. We should add a condition to the `deploy_preview.yaml` workflow to run only after a pull request event.

**Comment by mgrover1** [comment]
Great! Super exciting work @brian-rose . Should we week this open until that last point is resolved?

**Comment by brian-rose** [comment]
My perfectionist tendency says yes. I will take a look tomorrow.

**Comment by brian-rose** [comment]
@mgrover1 I think I'm done tinkering, finally!

The refactor solves the unwanted launching of the `deploy_preview.yaml` event, and will also make it easier to deploy these actions to many different cookbooks (and keep it all maintained!).

I'm going to start looking at migrating the workflows upstream to the Cookbook template.

**Comment by brian-rose** [comment]
Now that https://github.com/ProjectPythiaTutorials/cookbook-template/issues/9 is resolved, I think this work is actually done for now. I will open a new PR here that harmonizes this Cookbook with the latest version of the template.

**Comment by brian-rose** [comment]
This is now done with #42. Closing!
